jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -3614,6 +3656,41 @@ private static final void validateArgumentsOfCountedLoop(MethodHandle startHandl
 				/*[MSG "K065I", "The 2nd parameter type of the body handle with a non-void return type must be int rather than {0}"]*/
 				throw new IllegalArgumentException(com.ibm.oti.util.Msg.getString("K065I", bodyType.toMethodDescriptorString())); //$NON-NLS-1$
 			}
+			
+			bodyIterationVarLength = 2;
+		}
+
+		/* As a special case, if the body contributes only V and I types, with no
+		 * additional A types, then the internal parameter list is extended by the
+		 * argument types A... of the end handle. Adjust A count to correctly verify
+		 * parameters are effectively identical. Body will be altered outside of
+		 * verification.
+		 */
+		int bodyExternalParamLength = bodyParamLength - bodyIterationVarLength;
+		if (bodyParamLength == bodyIterationVarLength) {
+			bodyExternalParamLength = endHandle.type.parameterCount();
+		}
+		
+		/* Verify that handles are effectively identical and constrained by body. All we
+		 * need to do here is verify that no parameters have longer argument lists than
+		 * body. The generic loop will compare arguments to ensure all clauses are
+		 * effectively identical.
+		 */
```

Comment:
The lines just above this starting at 3669 make sure that the code you are looking at satisfies both cases by updating bodyExternalParamLength to the length of the end handle parameters if body has no external parameter list.

```
int bodyExternalParamLength = bodyParamLength - bodyIterationVarLength;
if (bodyParamLength == bodyIterationVarLength) {
	bodyExternalParamLength = endHandle.type.parameterCount();
}
```

Reply:
You merged the case 1) with case 2) to simplify the code but got reviewer confused with your logic.

1) if (bodyParamLength == bodyIterationVarLength) {
	bodyExternalParamLength = endHandle.type.parameterCount();
}  
<---- Given that the body has no external parameter, why reassigned it with endHandle.type.parameterCount()?  It seems the only reason is to reuse it for later comparison but logically doesn't make any sense.  The simple way is to rename "bodyExternalParamLength" to "externalParamLength" to avoid confusion.

2) Meanwhile, the error message in K0654 indicates we are comparing the content here (actually we don't).  If so, you need to print out the content of parameter types which incurs more unnecessary code here.  I would recommend to change it to something like:  The count of {0} 's parameters must be less than to the external parameters. ....
Better to print out the length of parameters in the error message to get users understand what happens to their application code.